core/vm/tosca_ct_integration.go

```diff
@@ -0,0 +1,165 @@
+package vm
+
+import (
+	"math/big"
+	"strings"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/holiman/uint256"
+)
+
+// Error
+var ErrStopToken = errStopToken
+
+// Gas table
+func MemoryGasCost(mem *Memory, wordSize uint64) (uint64, error) {
+	return memoryGasCost(mem, wordSize)
+}
+
+// Stack
+func NewStack() *Stack {
+	return newstack()
+}
+
+func (st *Stack) Len() int {
+	return st.len()
+}
+
+func (st *Stack) Push(d *uint256.Int) {
+	st.push(d)
+}
+
+// EVM
+func (evm *EVM) GetDepth() int {
+	return evm.depth
+}
+
+func (evm *EVM) SetDepth(depth int) {
+	evm.depth = depth
+}
+
+// Interpreter
+func (g *EVMInterpreter) SetLastCallReturnData(data []byte) {
+	g.returnData = data
+}
+
+func (g *EVMInterpreter) GetLastCallReturnData() []byte {
+	return g.returnData
+}
+
+func (g *EVMInterpreter) SetReadOnly(readOnly bool) {
+	g.readOnly = readOnly
+}
+
+func (g *EVMInterpreter) IsReadOnly() bool {
+	return g.readOnly
+}
+
+// CallContext Call interceptor
+
+// CallContext provides a basic interface for the EVM calling conventions. The EVM
+// depends on this context being implemented for doing subcalls and initialising new EVM contracts.
+type CallContextInterceptor interface {
+	// Call calls another contract.
+	Call(env *EVM, me ContractRef, addr common.Address, data []byte, gas, value *big.Int) ([]byte, uint64, error)
+	// CallCode takes another contracts code and execute within our own context
+	CallCode(env *EVM, me ContractRef, addr common.Address, data []byte, gas, value *big.Int) ([]byte, uint64, error)
+	// DelegateCall is same as CallCode except sender and value is propagated from parent to child scope
+	DelegateCall(env *EVM, me ContractRef, addr common.Address, data []byte, gas *big.Int) ([]byte, uint64, error)
+	// Create creates a new contract
+	Create(env *EVM, me ContractRef, data []byte, gas, value *big.Int) ([]byte, common.Address, uint64, error)
+
+	StaticCall(env *EVM, me ContractRef, addr common.Address, input []byte, gas *big.Int) ([]byte, uint64, error)
+	Create2(env *EVM, me ContractRef, code []byte, gas *big.Int, value *big.Int, salt *uint256.Int) ([]byte, common.Address, uint64, error)
+}
+
+// Interpreter interface
+// EVMInterpreter defines an interface for different interpreter implementations.
+type GethEVMInterpreter interface {
```

Comment:
if the idea behind this is that other implementations also interpret this interface, should it still be called `GethEVMInterpreter` or something a little more generic like `CtEvmInterpreter` or `ExternalInterpreter` ?

Reply:
this is the old interface of the geth interpreter, even without Tosca nor ct, still valid.

Reply:
It is the interface that the Geth EVM (not the interpreter) imposes on interpreter implementations. However, this could also be sufficiently covered by its package name.

If the name `Interpreter` is available (no collision in this namespace), I think it should be renamed.